src/inference/hmc.jl

```diff
@@ -529,7 +453,11 @@ gradient at `θ` for the model specified by `(vi, spl, model)`.
 """
 function gen_∂logπ∂θ(vi::VarInfo, spl::Sampler, model)
     function ∂logπ∂θ(x)
-        return gradient_logp(x, vi, model, spl)
```

Comment:
I believe this is a merge issue, we should use the simplified version.

Reply:
Can you double check if https://github.com/TuringLang/Turing.jl/pull/849/files is not overwritten.

Reply:
Everything in that PR made it over except for the progress stuff. The PR is designed in such a way that AHMC should overload a `callback` function to use at the interface level, and zero reporting functions occur within the sampler. The following two lines should be functionalized:

https://github.com/TuringLang/AdvancedHMC.jl/blob/691aa3deea03531fe0e36af5545d6671591c9089/src/sampler.jl#L70

https://github.com/TuringLang/AdvancedHMC.jl/blob/691aa3deea03531fe0e36af5545d6671591c9089/src/sampler.jl#L74

We can do this on the Turing side, but that's not very easy, because all of the things that AHMC reports with are local state variables. The better way to address this would be to move AHMC's `sample` function to mirror the interface setup and store all the reported variables in a mutable state, and then the `callback` function would just happen automatically in a unified way across the whole Turingverse without any idiosyncracies.

That was at least my interpretation of how the interface should work, but I'm happy to add it all back in. I personally think we should put in a little effort to standardize how progress is reported everywhere, which is why I added the `callback` stuff.  @yebai do you have an opinion here?

Reply:
Sorry, there were some things that didn't make it over, I'm fixing them.

Reply:
> The better way to address this would be to move AHMC's sample function to mirror the interface setup and store all the reported variables in a mutable state, and then the callback function would just happen automatically in a unified way across the whole Turingverse without any idiosyncracies.

We made a lot of effort to make sure things are immutable in AHMC side to avoid potential bugs. But it's pretty easy to wrap up it in the new mutable interface in Turing still. As I said I will give a try on integrating AHMC here.